File: python/euler/bin/p0042.py
# project euler: problem 42

from collections.abc import Iterable
from typing import TextIO

from euler.lib.util import is_triangular


def calc_scores(words: Iterable[str]) -> list[int]:
    a_to_z = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    tbl = dict(zip(a_to_z, range(1, len(a_to_z) + 1), strict=True))

    def score(word: str) -> int:
        return sum(map(lambda c: tbl[c], list(word)))

    return list(map(score, words))


def compute(fh: TextIO) -> str:
    keywords = [s.strip('"') for s in fh.read().split(',')]

    return str(len(list(filter(is_triangular, calc_scores(keywords)))))


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p042_words.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0042.ts
// project euler: problem 42

import { zip } from "std/collections/zip.ts";
import { assetData } from "../lib/asset.ts";
import { isTriangle, sum } from "../lib/math.ts";
import { range } from "../lib/util.ts";

const calcScore = (words: string[]): number[] => {
  const score = (word: string): number =>
    sum(word.split("").map((x) => tbl.get(x) as number));

  const AtoZ = range("A".charCodeAt(0), "Z".charCodeAt(0) + 1)
    .map((x) => String.fromCharCode(x));
  const tbl = new Map<string, number>(zip(AtoZ, range(1, AtoZ.length + 1)));

  return words.map((x) => score(x));
};

export const compute = (data: string): string => {
  const keywords = data.replaceAll('"', "").split(",");

  return String(
    calcScore(keywords).filter((x) => isTriangle(x) === true).length,
  );
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p042_words.txt"));
  return compute(data);
};
